share/tools/web_config/webconfig.py

```diff
@@ -1664,7 +1664,7 @@ def log_error(self, format, *args):
 
 # Try to find a suitable port
 PORT = 8000
-HOST = "::" if socket.has_ipv6 else "localhost"
+HOST = "::" if socket.has_dualstack_ipv6() else "localhost"
```

Comment:
I'm not sure this is correct (but then python's handling here is extremely confusing so it is possible I'm wrong).

It tells us if "the platform supports creating a SOCK_STREAM socket which can handle both AF_INET and AF_INET6 (IPv4 / IPv6) connections". That's not really something we're interested in.

What we *want* is to use one of the two, not both.

We're already trying to do this by checking if we get "EAFNOSUPPORT" - see below.

It appears what we need to do in addition to that is to check for a ValueError, as python might error out before:

https://github.com/python/cpython/blob/b430399d41fa88e9040cd055e55cf9211bf63c61/Lib/socket.py#L899-L900

(this is an extremely crappy error because instead of giving a specific value like "EAFNOSUPPORT" meaning "I do not support this address family" we now get a flat "This value ain't it, chief")

Reply:
Okay, first of all: Can you confirm that webconfig.py, as it currently stands in master (*not* 3.5.1, this isn't released yet) since 47ae6bc746c59020f1b3e341fd5a2445d7c497d3, does not work for you?

Because #3857 was closed because of that, and I did test at that time, and it appeared to work.

To test this you can also run `python $PWD/webconfig.py` without installing fish from master (it needs the absolute path tho, hence the $PWD).

>Even if the case might be very rare (especially for localhost) I suggest we replace my dualstack calls with a similar function like:

Please see https://github.com/fish-shell/fish-shell/blob/43d618e51f110dd370f112463ce8034eef889d26/share/tools/web_config/webconfig.py#L1674-L1682. That is almost exactly what we are doing now!

However: From reading the python code you linked, it seems python versions that have the "dualstack" methods can also throw a ValueError instead of the OSError with EAFNOSUPPORT. So all we would have to do is to *also* catch a ValueError *iff* that actually happens.

The issue with has_dualstack_ipv6() is that it is an additional method that I'm not sure does the actual right thing, and that it appears to have been introduced in python 3.7, while the rest of this works with older versions as well (you would be surprised at how many bug reports we get by people using new fish on otherwise ancient systems - RHEL exists, and people do *not* upgrade quickly).

In fact whether it is "dualstack" isn't something we really care about - we only want to know if we can use an ipv6 or an ipv4 address. We don't need a socket that can do both, which is what the "dualstack" thing is apparently about.

>Also I would like to remove the "localhost" string

Sure.

Reply:
Okay yeah, the master is working for me. There could be some improvements but it doesn't really matter.

Also about the ValueError, I don't think you should worry about it since it needs to be enabled [in](https://github.com/python/cpython/blob/b430399d41fa88e9040cd055e55cf9211bf63c61/Lib/socket.py#L898) which unless I'm missing something, its not.

I agree with the rest, you can close this PR. Sorry for occupying you.